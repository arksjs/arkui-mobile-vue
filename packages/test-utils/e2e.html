<script src="../../node_modules/vue/dist/vue.global.js"></script>
<script src="../../dist/index.js"></script>
<link rel="stylesheet" href="../../dist/index.css" />

<div id="app"></div>
