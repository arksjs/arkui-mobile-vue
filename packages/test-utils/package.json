{
  "name": "@arksjs/test-utils",
  "version": "0.0.1",
  "description": "arkui test-utils",
  "main": "dist/index.js",
  "private": true
}
