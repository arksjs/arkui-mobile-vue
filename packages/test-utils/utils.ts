export const timeout = (n: number) => new Promise(r => setTimeout(r, n))
