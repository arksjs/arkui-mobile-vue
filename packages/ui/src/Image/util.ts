import { camelCase2KebabCase, getNumber, getEnumsValue, type CSSProperties } from '../helpers'
import type { Mode } from './types'

export const MODE_NAMES: Mode[] = [
  'scaleToFill',
  'aspectFit',
  'aspectFill',
  'widthFix',
  'top',
  'bottom',
  'left',
  'right',
  'top left',
  'top right',
  'bottom left',
  'bottom right'
]

export const getClasses = (autoHeight?: boolean) => [
  'ta-image',
  {
    'auto-height': !!autoHeight
  }
]

export const getImgClasses = (mode?: Mode, autoHeight?: boolean) => [
  'ta-image_img',
  'mode--' +
    camelCase2KebabCase(autoHeight ? 'widthFix' : getEnumsValue(MODE_NAMES, mode)).replace(
      /\s+/g,
      '-'
    )
]

export const getRatioStyles = (aspectRatio?: number | string) => {
  return {
    paddingTop: getNumber(aspectRatio) * 100 + '%'
  } as CSSProperties
}
