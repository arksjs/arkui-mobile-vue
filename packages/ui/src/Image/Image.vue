<template>
  <div :class="classes" ref="root">
    <span
      v-if="!autoHeight && aspectRatio != null"
      class="ta-image_ratio"
      :style="ratioStyles"
    ></span>
    <i class="ta-image_loading" v-if="loading">
      <Icon :icon="loadingIcon" :size="iconSize" />
    </i>
    <i class="ta-image_error" v-if="error">
      <Icon :icon="errorIcon" :size="iconSize" />
    </i>
    <img v-if="currentSrc" :class="imgClasses" :src="currentSrc" @dragstart="onDrag" />
  </div>
</template>

<script lang="ts">
import {
  defineComponent,
  onMounted,
  ref,
  watch,
  onBeforeUnmount,
  computed,
  shallowRef,
  type PropType
} from 'vue'
import { Icon } from '../Icon'
import {
  addLazyQueue,
  loadNow,
  removeComponentFromLazy,
  withCheckInView
} from '../Image/load-image'
import {
  createEnumsValidator,
  emitErrorValidator,
  iconValidator,
  isString,
  type OnError,
  type PropsToEmits
} from '../helpers'
import type { ImageEmits, LoadedResource, Mode } from './types'
import ImageOutlined from '../Icon/icons/ImageOutlined'
import ImageBreakOutlined from '../Icon/icons/ImageBreakOutlined'
import type { IconData } from '../Icon/types'
import { getClasses, getImgClasses, getRatioStyles, MODE_NAMES } from './util'
import { useException } from '../hooks'

export default defineComponent({
  name: 'ta-image',
  components: { Icon },
  props: {
    // 图片资源地址
    src: {
      type: String,
      default: ''
    },
    mode: {
      type: String as PropType<Mode>,
      validator: createEnumsValidator(MODE_NAMES)
    },
    // 图片懒加载，在即将进入一定范围(preload=1.3)时才开始加载
    lazyLoad: {
      type: Boolean,
      default: false
    },
    // 自适应正方形
    aspectRatio: {
      type: [Number, String]
    },
    // 允许拖拽
    draggable: {
      type: Boolean,
      default: true
    },
    // 加载图标名称
    loadingIcon: {
      type: [String, Object] as PropType<IconData>,
      validator: iconValidator,
      default: ImageOutlined
    },
    // 加载失败图标名称
    errorIcon: {
      type: [String, Object] as PropType<IconData>,
      validator: iconValidator,
      default: ImageBreakOutlined
    },
    iconSize: {
      type: [Number, String]
    },
    // 自动计算高度
    autoHeight: {
      type: Boolean,
      default: false
    }
  },
  emits: {
    load: payload => payload && payload.width > 0 && payload.height > 0 && isString(payload.src),
    error: emitErrorValidator
  } as PropsToEmits<ImageEmits>,
  setup(props, { emit }) {
    const { createException } = useException()
    const loading = ref(true)
    const error = ref(false)
    const root = shallowRef<HTMLElement | null>(null)
    const currentSrc = ref<string | null>(null)
    const uid = Symbol()

    function load(src: string) {
      ;(props.lazyLoad ? addLazyQueue : loadNow)({
        src,
        uid,
        checkInView: withCheckInView(root.value as HTMLElement),
        onLoad,
        onError
      })
    }

    const onLoad = (res: LoadedResource) => {
      if (res.src === props.src) {
        // 防止多次变更图片导致的图片不正确
        loading.value = false
        error.value = false
        currentSrc.value = res.src
      }

      emit('load', {
        width: res.naturalWidth,
        height: res.naturalHeight,
        src: res.src
      })
    }

    const onError: OnError = e => {
      loading.value = false
      error.value = true

      emit('error', createException(e))
    }

    function onDrag(e: Event) {
      if (!props.draggable) {
        e.preventDefault()
      }
    }

    onMounted(() => props.src && load(props.src))

    onBeforeUnmount(() => removeComponentFromLazy(uid))

    watch(
      () => props.src,
      val => load(val)
    )

    const classes = computed(() => getClasses(props.autoHeight))
    const imgClasses = computed(() => getImgClasses(props.mode, props.autoHeight))
    const ratioStyles = computed(() => getRatioStyles(props.aspectRatio))

    return {
      currentSrc,
      classes,
      imgClasses,
      ratioStyles,
      loading,
      error,
      root,
      onDrag
    }
  }
})
</script>
