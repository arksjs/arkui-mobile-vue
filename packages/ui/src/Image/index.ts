import { withInstall } from '../helpers/with-install'
import Image from './Image.vue'

export { Image }
export const install = withInstall(Image)
export default Image
