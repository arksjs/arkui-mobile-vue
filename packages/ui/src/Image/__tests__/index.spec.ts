import { describe, test, expect } from 'vitest'
import { mount, shallowMount } from '@vue/test-utils'
import { Image } from '@/Image'
import { nextTick } from 'vue'

const imgUrl = 'https://cdn.fox2.cn/vfox/swiper/center-2.jpg'

describe('Image', () => {
  test('snapshot', () => {
    const wrapper = shallowMount(Image, {})
    expect(wrapper.html()).toMatchSnapshot()
  })

  test.concurrent('mode', async () => {
    const wrapper = mount(Image, {})
    wrapper.vm.currentSrc = imgUrl

    await nextTick()
    const $img = wrapper.find('.ta-image_img')

    expect($img.classes()).toContain('mode--scale-to-fill')

    await wrapper.setProps({
      mode: 'aspectFit'
    })
    expect($img.classes()).toContain('mode--aspect-fit')

    await wrapper.setProps({
      mode: 'aspectFill'
    })
    expect($img.classes()).toContain('mode--aspect-fill')

    await wrapper.setProps({
      mode: 'widthFix'
    })
    expect($img.classes()).toContain('mode--width-fix')

    await wrapper.setProps({
      mode: 'top'
    })
    expect($img.classes()).toContain('mode--top')

    await wrapper.setProps({
      mode: 'bottom'
    })
    expect($img.classes()).toContain('mode--bottom')

    await wrapper.setProps({
      mode: 'left'
    })
    expect($img.classes()).toContain('mode--left')

    await wrapper.setProps({
      mode: 'right'
    })
    expect($img.classes()).toContain('mode--right')

    await wrapper.setProps({
      mode: 'top left'
    })
    expect($img.classes()).toContain('mode--top-left')

    await wrapper.setProps({
      mode: 'top right'
    })
    expect($img.classes()).toContain('mode--top-right')

    await wrapper.setProps({
      mode: 'bottom left'
    })
    expect($img.classes()).toContain('mode--bottom-left')

    await wrapper.setProps({
      mode: 'bottom right'
    })
    expect($img.classes()).toContain('mode--bottom-right')
  })

  test.concurrent('aspectRatio', async () => {
    const wrapper = mount(Image, {
      props: {
        aspectRatio: 1
      }
    })
    const $el = wrapper.find('.ta-image_ratio').element as HTMLElement
    expect($el.style.paddingTop).toContain('100%')
  })
})
