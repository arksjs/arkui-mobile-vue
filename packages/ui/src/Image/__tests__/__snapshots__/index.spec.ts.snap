// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Image > snapshot 1`] = `
"<div class=\\"ta-image\\">
  <!--v-if--><i class=\\"ta-image_loading\\">
    <icon-stub icon=\\"[object Object]\\" spin=\\"false\\" size=\\"0\\" width=\\"0\\" height=\\"0\\"></icon-stub>
  </i>
  <!--v-if-->
  <!--v-if-->
</div>"
`;
